From 72aadd897b2f07f5191859ae1b94d51425828fa3 Mon Sep 17 00:00:00 2001
From: Joe Heck <heckj@mac.com>
Date: Fri, 1 Jun 2012 21:51:48 +0000
Subject: [PATCH] Tolerance when parsing null connection info in BDM

Resolves bug 1007615.

Defensive coding against attempt to decode None as a JSON-encoding
of connection info in an instance block device mapping.

(cherry picked from commit 7d57cc16a6ab3b286e4fd3a479e5b40160980304)

Change-Id: If7afa4fb030b3c53a0b80737ff792e42cc4d3101
---
 nova/compute/manager.py |   13 +++++++++----
 1 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/nova/compute/manager.py b/nova/compute/manager.py
index ca51fc5..2d1745b 100644
--- a/nova/compute/manager.py
+++ b/nova/compute/manager.py
@@ -659,10 +659,15 @@ class ComputeManager(manager.SchedulerDependentManager):
         bdms = self._get_instance_volume_bdms(context, instance_id)
         block_device_mapping = []
         for bdm in bdms:
-            cinfo = utils.loads(bdm['connection_info'])
-            block_device_mapping.append({'connection_info': cinfo,
-                                         'mount_device':
-                                         bdm['device_name']})
+            try:
+                cinfo = utils.loads(bdm['connection_info'])
+                block_device_mapping.append({'connection_info': cinfo,
+                                             'mount_device':
+                                             bdm['device_name']})
+            except TypeError:
+                # if the block_device_mapping has no value in connection_info
+                # (returned as None), don't include in the mapping
+                pass
         # NOTE(vish): The mapping is passed in so the driver can disconnect
         #             from remote volumes if necessary
         return {'block_device_mapping': block_device_mapping}
